html/semantics/forms/constraints/support/validator.js

```diff
@@ -167,14 +167,14 @@ var validator = {
     test(function () {
       self.pre_check(ctl, "customError");
       self.iterate_over(ctl, data).forEach(function(val) {
-        const {ctl, data, condStr} = val;
-        if (data.expected) {
+        const {ctl, expected, condStr} = val;
+        if (expected) {
           assert_true(
               ctl.validity.customError,
               'The validity.customError attribute should be true' + condStr);
           // validationMessage returns the empty string if ctl is barred from
-          // constraint validation, which happens if ctl is disabled.
-          if (ctl.disabled) {
+          // constraint validation, which happens if ctl is disabled or readOnly.
+          if (ctl.disabled || ctl.readOnly) {
```

Comment:
Why ctl.readOnly check is necessary?

Reply:
[Readonly elements are barred from constraint validation so validationMessage should be empty](https://html.spec.whatwg.org/multipage/input.html#attr-input-readonly). There was a typo in iterate_over - `readonly` instead of `readOnly`:

https://github.com/web-platform-tests/wpt/blob/91bd700f07b3fe67d5c582e4ebe954516cd46875/html/semantics/forms/constraints/support/validator.js#L403

As a result elements weren't really readonly so the bug didn't appear. I fixed it in this PR, sorry I forgot to mention that.

Reply:
Ack.  `readOnly` affects `validationMessage` according to the standard.